### PR TITLE
Add a fallback argument to handle unreachable file

### PR DESCRIPTION
**Pull request by lo-tp**
#180
Imagine, we are debugging a react app with react-router. We click
a link sending us to another url. That's ok since all the routing
stuff are handled in the front end. Then we make some modification
to the html and want to refresh the whole page. Now problem arises,
the url is now unreachable in the back end because we have changed
it so that the server has no idea what file to serve.

In this case, we would like the dev server to serve one similar
html no matter what the url is( the root html). That would be the behavior
when fallBack argument is set to true after this commit is merged.

**What kind of change does this PR introduce?**
I added a new feature.

**Did you add tests for your changes?**
Yes

**Does this PR introduce a breaking change?**
No. The default behavior is untouched if the fallBack argument is
not set to true.

### CODE REVIEW

File: middleware.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ module.exports = function(compiler, options) {
 	var shared = Shared(context);
 
 
-	// The middleware function
```
Comment (shellscape) [inline middleware.js]: please don't change whitespace formatting :)

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "webpack-dev-middleware",
-  "version": "1.10.1",
+  "version": "1.10.2",
```
Comment (shellscape) [inline package.json]: please don't update package versions, we'll do that on our end for publish

File: test/API.test.js [hunk 3]
```diff
@@ -1,5 +1,6 @@
 var middleware = require("../middleware");
 var should = require("should");
+require("mocha-sinon");
```
Comment (shellscape) [inline test/API.test.js]: please move this to line 1 and add a line break between it and `var middleware...`

**Comment by shellscape** [comment]
@lo-tp thanks for cleaning up those spots. looks like there's a few conflicts to resolve. I'd recommend rebasing

**Comment by lo-tp** [comment]
@shellscape Done rebasing.

**Comment by lo-tp** [comment]
Hi, @shellscape , I have done rebasing, is there any other thing I can do to speed up the merging process?

File: middleware.js (hunk 1 above)
Comment (shellscape) [inline middleware.js]: @lo-tp please revert the whitespace changes

File: middleware.js [hunk 4]
```diff
@@ -40,6 +40,9 @@ module.exports = function(compiler, options) {
 		if(filename === false) return goNext();
 
 
+		if(!context.fs.existsSync(filename) && context.options.fallBack) {
+			filename = getFilenameFromUrl(context.options.publicPath, context.compiler.outputPath,context.options.publicPath);
```
Comment (shellscape) [inline middleware.js]: inconsistent spaces after commas. it's a nitpick but it's good to be consistent

**Comment by shellscape** [comment]
@lo-tp looks like your editor is setup for some specific whitespace settings and it may be automatically changing things. please revert the whitespace changes and clean up spacing between commas (i'll look at improving the linting in the future). 

other than that the feature looks good. comparing for a truthy value of `options.fallback` will allow for furthering this feature in the future if needed.

**Comment by lo-tp** [comment]
@shellscape . My fault, I'll make sure that all the modifications I made are consistent with the eslint configuration in the code base.

**Comment by shellscape** [comment]
@lo-tp thanks for making those changes. I was discussing the PR offline and was asked about the  option in webpack-dev-server. https://webpack.github.io/docs/webpack-dev-server.html#the-historyapifallback-option. So we have to ask - can you use this option instead of the feature you've implemented in this PR? We want to make sure we're not duplicating functionality.

(Folks that don't use webpack-dev-server (who are using express, koa, etc) Can use https://github.com/bripkens/connect-history-api-fallback, which is the same middleware that webpack-dev-server uses for the option.)

**Comment by lo-tp** [comment]
@shellscape , I am so happy to see that there is an existing solution to my problem.
With https://github.com/bripkens/connect-history-api-fallback, every thing works so smoothly.
Think you for telling me this package.

**Comment by shellscape** [comment]
Wonderful! Thanks for all of your work on this PR, and I'm sorry I wasn't able to point you at that module sooner.

**Comment by kenjura** [comment]
I don't understand how connect-history-api-fallback is a solution to this problem. Perhaps I'm not implementing it correctly. Here is the scenario:

In local dev, I start my project by running index.js. This activates an Express backend, then kicks in webpack-dev-middleware to compile the frontend and serve it.

Once ready, requests may go to myapp.com/, but also myapp.com/foo/bar or myapp.com/foo/123

In the former case, webpack-dev-middleware serves the frontend bundle as expected, and history API changes work as expected. However, the local dev server should also respond to the latter two URLs, serving the exact same frontend bundle, as would happen with the fallBack feature in this PR.

The problem with connect-history-api-fallback seems to be that it is assuming the file exists on the local file system. By design, webpack-dev-middleware creates an in-memory file system. The index.html and all bundled assets reside there--not on the file system--and therecore connect-history-api-fallback cannot serve them.

In my experience, SPAs typically operate from a single root HTML file with all necessary JS to render all possible routes, and although they use a frontend router and the History API to navigate within a session, their web servers should also respond to deep links exactly the same as if the user started from the index page and navigated using History.

In order for that to work with webpack-dev-middleware, it would seem the feature in this PR is required.

Am I not understanding something? Is there some other solution that involves webpack-dev-middleware? I was previously using webpack-dev-server plus a separate process for the backend, but I thought the point of this middleware was that I could unify the two.

**Comment by AlexKvazos** [comment]
@kenjura Mentioned something very important. Currently I see no way of using webpack-dev-middleware to serve SPA without having to request the base url every refresh.

**Pull request closed without merging**